mixnet/src/fragment.rs

```diff
@@ -1,22 +1,249 @@
+            body: value[FRAGMENT_HEADER_SIZE..].to_vec(),
+        })
     }
 }
 
-pub struct MessageReconstructor {}
+type FragmentSetId = Uuid;
+type FragmentId = u8;
+const MAX_NUM_FRAGMENTS: usize = FragmentId::MAX as usize + 1;
+const FRAGMENT_HEADER_SIZE: usize = 18;
+
+#[derive(PartialEq, Eq, Debug, Clone)]
+struct FragmentHeader {
+    set_id: FragmentSetId,
+    total_fragments: FragmentId,
+    fragment_id: FragmentId,
+}
+
+impl FragmentHeader {
+    fn bytes(&self) -> [u8; FRAGMENT_HEADER_SIZE] {
+        let mut out = [0u8; FRAGMENT_HEADER_SIZE];
+        out[0..16].copy_from_slice(self.set_id.as_bytes());
+        out[16] = self.total_fragments;
+        out[17] = self.fragment_id;
+        out
+    }
+
+    fn from_bytes(value: &[u8]) -> Result<Self, MixnetError> {
+        if value.len() != FRAGMENT_HEADER_SIZE {
+            return Err(MixnetError::InvalidFragmentHeader);
+        }
+
+        Ok(Self {
+            set_id: Uuid::from_slice(&value[0..16])?,
+            total_fragments: value[16],
+            fragment_id: value[17],
+        })
+    }
+}
+
+pub struct MessageReconstructor {
+    fragment_sets: HashMap<FragmentSetId, FragmentSetReconstructor>,
+}
 
 impl MessageReconstructor {
     pub fn new() -> Self {
-        todo!()
+        Self {
+            fragment_sets: HashMap::new(),
+        }
     }
 
     pub fn add(&mut self, fragment: Fragment) -> Option<Vec<u8>> {
-        todo!()
+        let set_id = fragment.header.set_id;
+        self.fragment_sets
+            .entry(set_id)
+            .or_insert(FragmentSetReconstructor::new(
+                fragment.header.total_fragments,
+            ))
+            .add(fragment)
+            .map(|msg| {
+                // A message has been reconstructed completely from the fragment set.
+                // Delete the fragment set from the reconstructor.
+                self.fragment_sets.remove(&set_id);
+                msg
+            })
+    }
+}
+
+struct FragmentSetReconstructor {
+    total_fragments: FragmentId,
+    fragments: HashMap<FragmentId, Fragment>,
+    // For mem optimization, accumulates the expected message size
+    // whenever a new fragment is added to the `fragments`.
+    message_size: usize,
+}
+
+impl FragmentSetReconstructor {
+    fn new(total_fragments: FragmentId) -> Self {
+        Self {
+            total_fragments,
+            fragments: HashMap::new(),
+            message_size: 0,
+        }
+    }
+
+    fn add(&mut self, fragment: Fragment) -> Option<Vec<u8>> {
+        self.message_size += fragment.body.len();
+        if let Some(old_fragment) = self.fragments.insert(fragment.header.fragment_id, fragment) {
+            // In the case when a new fragment replaces the old one, adjust the `meesage_size`.
+            // e.g. The same fragment has been received multiple times.
+            self.message_size -= old_fragment.body.len();
+        }
+
+        self.try_build_message()
+    }
```

Comment:
This feels more like a builder pattern. I would rather have this return a `&mut self` so calls to `try_build_message` are explicit and `add` can be chained.

```rust
builder.add(f1).add(f2).try_build_message()
```

Reply:
I agree that `add()` could be confusing. For `FragmentSetReconstructor`, yes, it's more clear to call `try_build_message()` explicitly. But, for `MessageReconstructor` (which is an end-user interface), I think it's better to make users to call only one function which adds fragment, tries to reconstruct the message, and cleans up internal data structures if the message has been reconstructed completely, because the `MessageReconstructor` contains multiple `FragmentSetReconstructor`. 

So, I refactored code like 8b35585c5f84c89bfb86a69b4c29a25a8fa41d48. I renamed the function name `MessageReconstructor.add()` more clearly. And, I changed it to call the functions of  `FragmentSetReconstructor` more explicitly. There's no case which we need to chain `add`s but I agree that it's better to call functions explictly.

Reply:
Cool, best of both worlds!